File: 2378-sender-with-largest-word-count/sender-with-largest-word-count.java
import java.util.*;

class Solution {
    public String largestWordCount(String[] messages, String[] senders) {
        
        int maxCount = 0;
        String ans = "";
        // Map<String, List<String>> wordCountMap = new HashMap<>();
        Map<String, Integer> wordCountMap = new HashMap<>();
        
        int n = messages.length;

        for(int i = 0; i < n; i ++) {
            // List<String> msg_list = Arrays.asList(messages[i].split(" "));
            // wordCountMap.computeIfAbsent(senders[i], k -> new ArrayList<>())
            // .addAll(msg_list);
            // int totalWords = wordCountMap.getOrDefault(senders[i], new ArrayList<>()).size();

            int currWords = messages[i].split(" ").length;
            int totalWords = wordCountMap.merge(senders[i], currWords, Integer::sum);

            if (maxCount < totalWords || (maxCount == totalWords && senders[i].compareTo(ans) > 0)) {
                maxCount = totalWords;
                ans = senders[i];
            }
        }      
        return ans;
    }    
}

File: 2378-sender-with-largest-word-count/sender-with-largest-word-count.py
from collections import defaultdict
class Solution:
    def largestWordCount(self, messages: List[str], senders: List[str]) -> str:
        # case- sensitive
        # wordcount ==  sort(count, key)
        # alice = 3 and Alice = 3

        maxCount = 0
        ans = ""
        wordCountMap = defaultdict(list)

        for msg, sender in zip(messages, senders):
            msg_list = msg.split(" ")
            wordCountMap[sender].extend(msg_list)
            
            currCount = len(wordCountMap[sender])
            if maxCount < currCount:
                maxCount = currCount
                ans = sender
            elif maxCount == currCount:
                ans = max(ans, sender)
        return ans


